optimizer/plan/predicate_push_down.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plan
+
+import (
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/ast"
+)
+
+func addFilter(p Plan, child Plan, conditions []ast.ExprNode) error {
+	filter := &Filter{Conditions: conditions}
+	return InsertPlan(child, p, filter)
+}
+
+// columnSubstituor substitutes the columns in filter to expressions in select fields.
+// e.g. Filter(a < 10) <- SelectFields(1 as a) ==> Select(1 as a) <- Filter(1 < 10).
+type columnSubstitutor struct {
+	fields []*ast.ResultField
+}
+
+func (cl *columnSubstitutor) Enter(inNode ast.Node) (node ast.Node, ok bool) {
+	switch v := inNode.(type) {
+	case *ast.ColumnNameExpr:
+		match := false
+		for _, field := range cl.fields {
+			if field == v.Refer {
+				switch field.Expr.(type) {
+				case *ast.ColumnNameExpr:
+				case *ast.ValueExpr:
+				default:
+					return nil, false
+				}
+				match = true
+			}
+		}
+		return node, match
+	}
+	return inNode, true
+}
+
+func (cl *columnSubstitutor) Leave(inNode ast.Node) (node ast.Node, ok bool) {
+	switch v := inNode.(type) {
+	case *ast.ColumnNameExpr:
+		for _, field := range cl.fields {
+			if v.Refer == field {
+				return field.Expr, true
+			}
+		}
+	}
+	return inNode, true
+}
+
+// PredicatePushDown applies predicate push down to all kinds of plans, except aggregation and union.
+func PredicatePushDown(p Plan, predicates []ast.ExprNode) (ret []ast.ExprNode, err error) {
+	if len(p.GetChildren()) == 0 {
+		return predicates, nil
+	}
+	switch v := p.(type) {
+	case *TableScan:
```

Comment:
If p has no children, then the type of p would never be `TableScan` as `TableScan` has no children.